src/main/java/org/wildfly/naming/client/remote/RemoteNamingService.java

```diff
@@ -0,0 +1,133 @@
+                    public void handleMessage(Channel channel, MessageInputStream message) {
+                        try (MessageInputStream mis = message) {
+                            byte[] namingHeader = new byte[6];
+                            mis.read(namingHeader);
+                            if (! Arrays.equals(namingHeader, ProtocolUtils.NAMING_BYTES)) {
+                                throw Messages.log.invalidHeader();
+                            }
+                            int version = mis.readUnsignedByte();
+                            boolean versionSupported = false;
+                            for (int supportedProtocolVersion : SUPPORTED_PROTOCOL_VERSIONS) {
+                                if (version == supportedProtocolVersion) {
+                                    versionSupported = true;
+                                    break;
+                                }
+                            }
+                            if (! versionSupported) {
+                                throw Messages.log.unsupportedProtocolVersion(version);
+                            }
+                            final RemoteServerTransport remoteServerTransport = new RemoteServerTransport(channel, version, messageTracker, RemoteNamingService.this);
+                            remoteServerTransport.start();
+                        } catch (IOException | CommunicationException e) {
+                            Messages.log.failedToDetermineClientVersion(e);
+                        }
+                    }
+
+                    public void handleError(final Channel channel, final IOException error) {
+                        try {
+                            channel.close();
+                        } catch (IOException ignored) {
+                        }
+                    }
+
+                    public void handleEnd(final Channel channel) {
+                        try {
+                            channel.close();
+                        } catch (IOException ignored) {
+                        }
+                    }
+                });
+                // Send greeting message
+                try (MessageOutputStream mos = messageTracker.openMessage()) {
+                    mos.write(ProtocolUtils.NAMING_BYTES);
+                    mos.writeByte(SUPPORTED_PROTOCOL_VERSIONS.length);
+                    for (byte version : SUPPORTED_PROTOCOL_VERSIONS) {
+                        mos.writeByte(version);
+                    }
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    IoUtils.safeClose(channel);
+                }
+            } catch (IOException e) {
+                Messages.log.failedToSendHeader(e);
+                IoUtils.safeClose(channel);
+            }
+        }
+
+        public void registrationTerminated() {
+        }
+    }
+
+    public Context getLocalContext() {
+        return localContext;
```

Comment:
Actually this would only come into play when you are handling concurrent requests, which you are not (unless you move the receiveMessage(this) call up to the top of your message listener).